pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -356,34 +360,72 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
 		}
 	}
 
-	return validatePipelineVariables(tasks, "params", parameterNames, arrayParameterNames)
+	return validatePipelineParametersVariables(tasks, "params", parameterNames, arrayParameterNames)
```

Comment:
this would be misleading for invalid params in when expressions, would produce the following error for the test case in [pipeline_validation_test.go](https://github.com/tektoncd/pipeline/pull/3135/files#diff-d1d41b803d92fc5570cfcefb5e355959R865) in which `baz` is not defined as a pipeline param but used in `when` expression:

```
validation failed: non-existent variable in "$(params.baz)" for task parameter whenInput[$(params.baz)]: pipelinespec.params.whenInput[$(params.baz)] 
```

This function is part of validating Pipeline Params and the prefix `params` is used for validating params and also when expression. Even if this prefix is fixed,[ the nested function](https://github.com/tektoncd/pipeline/pull/3135/files#diff-5b096c7641765902d79ee2d5e5f7d213R430) has it hardcoded as `task parameter` and `pipelineSpec.params` which is wouldn't be valid for `pipelineSpec.tasks.when`.

This calls for a huge refactoring of pipeline validation 😢   Again not blocking this PR here but worth following up in separate PR.

Reply:
good catch, made the location name and path variables passed into the nested functions